### PR TITLE
Multi realm installs - Allow for modification of enabled realms…

### PR DESCRIPTION
**Pull request by tpetrus**
Integrate selection of desired realms to be enabled in XDMoD instance and allow all tests to pass based on which realms are enabled.
## Description

- Environment variable XDMOD_REALMS now used to set desired realms in a comma-separated string style
- Modified bootstrap among other files to only ingest proper items required for the realms desired
- Made Job Viewer tab disabled if the jobs realm isn't enabled, and suppressed error messages corresponding to JobViewer not finding the jobs realm (if it's not enabled)
- Updated all tests to properly work with whatever combination of realms that are enabled (This included splitting up the proper test artifacts by realm)

## Motivation and Context

It allows for the selection of realms to be enabled, where before there was no way to do so.

## Tests performed

I ran all tests on every single combination (and order) of Jobs, Cloud, and Storage realms possible, using both the fresh_install and upgrade test modes, and all on fresh XDMoD instances.

I ran the same script commands that Shippable currently runs to do its tests of XDMoD.  This included setting up a new XDMoD instance with the desired realms, and running regression, integration, component, UI, and all SSO tests on that instance.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: classes/Models/Services/Users.php [hunk 1]
```diff
@@ -392,7 +392,7 @@ public static function demoteUserFromCenterStaff(XDUser $user, $centerId)
 
         // If this user has no more center staff centers then remove the center
         // staff acl.
-        if (count(array_diff(array_values($currentCenters), array((string)$centerId))) === 0) {
+        if (empty($currentCenters) || count(array_diff(array_values($currentCenters), array((string)$centerId))) === 0) {
```
Comment (ryanrath) [inline classes/Models/Services/Users.php]: You should be able to replace this with: 
```php
if (empty($currentCenters)) {
        $currentCenters = array($user->getOrganizationID());
}

// If this user has no more center staff centers then remove the center
// staff acl.
if (count(array_diff(array_values($currentCenters), array((string)$centerId))) === 0) {
```

File: html/gui/js/modules/job_viewer/JobViewer.js [hunk 2]
```diff
@@ -128,6 +128,11 @@ XDMoD.Module.JobViewer = Ext.extend(XDMoD.PortalModule, {
         this.on('role_selection_change', this.noOpt);
         this.on('duration_change', this.noOpt);
 
+        // Disable Job Viewer tab if there is no Jobs realm enabled
```
Comment (ryanrath) [inline html/gui/js/modules/job_viewer/JobViewer.js]: This shouldn't be needed. We can now modify roles.d/jobs.json to determine which tabs are included. 
Remove from roles.json#roles#default#permitted_modules: 
```json 
{
                    "name": "job_viewer",
                    "title": "Job Viewer",
                    "position": 5000,
                    "javascriptClass": "XDMoD.Module.JobViewer",
                    "javascriptReference": "CCR.xdmod.ui.jobViewer",
                    "tooltip": "View detailed job-level metrics",
                    "userManualSectionName": "Job Viewer"
                },
```
and move it to: roles.d/jobs.json#roles#default#permitted_modules ( You'll need to create the key )

File: tests/ci/bootstrap.sh [hunk 3]
```diff
@@ -23,28 +28,109 @@ then
     mysql -e "CREATE USER 'root'@'gateway' IDENTIFIED BY '';
     GRANT ALL PRIVILEGES ON *.* TO 'root'@'gateway' WITH GRANT OPTION;
     FLUSH PRIVILEGES;"
-    expect $BASEDIR/scripts/xdmod-setup.tcl | col -b
+
+    #TODO: Replace diff files with hard fixes
+    if [[ $XDMOD_REALMS != *"jobs"* ]];
+    then
+        # -- Remove jobs realm
+        patch -up1 --directory=/usr/share/xdmod/ < $SRCDIR/xdmod/tests/ci/diff/nojobs-usr-share-xdmod.diff
```
Comment (ryanrath) [inline tests/ci/bootstrap.sh]: Shouldn't need to do these patches.

File: tests/ci/bootstrap.sh [hunk 4]
```diff
@@ -23,28 +28,109 @@ then
...
-    expect $BASEDIR/scripts/xdmod-upgrade.tcl | col -b
+
+    #TODO: Replace diff files with hard fixes
+    if [[ $XDMOD_REALMS != *"jobs"* ]];
+    then
+        # -- Remove jobs realm
+        patch -up1 --directory=/usr/share/xdmod/ < $SRCDIR/xdmod/tests/ci/diff/nojobs-usr-share-xdmod.diff
+        #patch -up2 --directory=/etc/xdmod/ < $SRCDIR/xdmod/tests/ci/diff/nojobs-configuration.diff
+        # -- Remove Job Viewer Listener (When Job Viewer tab has no data)
+        patch -up1 --directory=/usr/share/xdmod/ < $SRCDIR/xdmod/tests/ci/diff/rm_job_viewer_listener.diff
```
Comment (ryanrath) [inline tests/ci/bootstrap.sh]: If Job Viewer can not be shown then this patch shouldn't be needed anymore.

File: tests/component/lib/AclsTest.php [hunk 5]
```diff
@@ -54,6 +55,16 @@ class AclsTest extends BaseTest
         '_disable_menu'
     );
 
+    protected function setUp()
```
Comment (ryanrath) [inline tests/component/lib/AclsTest.php]: Add this to `BaseTest` and then you can just check `$this->xdmod_realms` in all the tests.

File: tests/component/lib/QueryTest.php [hunk 6]
```diff
@@ -16,6 +17,18 @@ class AggregateTest extends BaseTest
      */
     public function testGetDurrationResult($period, $start, $end, $groupby, $expected)
     {
+        //Get current realms enabled
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/component/lib/QueryTest.php]: *See comment in `AclsTest`

File: tests/component/lib/SlurmHelperTest.php [hunk 7]
```diff
@@ -48,6 +49,18 @@ public function testSlurmHelper(
         $outputRegex,
         $exitStatus
     ) {
+        //Get current realms enabled
```
Comment (ryanrath) [inline tests/component/lib/SlurmHelperTest.php]: *See comment in `AclsTest`

File: tests/component/lib/XDUserTest.php [hunk 8]
```diff
@@ -21,6 +22,17 @@ class XDUserTest extends BaseTest
 {
     private static $users = array();
 
+    protected function setUp()
```
Comment (ryanrath) [inline tests/component/lib/XDUserTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/BaseUserAdminTest.php [hunk 9]
```diff
@@ -58,9 +59,15 @@ abstract class BaseUserAdminTest extends \PHPUnit_Framework_TestCase
 
     protected function setUp()
     {
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Controllers/BaseUserAdminTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/MetricExplorerTest.php [hunk 10]
```diff
@@ -1,12 +1,20 @@
...
 
 namespace IntegrationTests\Controllers;
+use Models\Services\Realms;
 
 class MetricExplorerTest extends \PHPUnit_Framework_TestCase
 {
     protected function setUp()
     {
         $this->helper = new \TestHarness\XdmodTestHelper();
+        //Get current realms enabled
```
Comment (ryanrath) [inline tests/integration/lib/Controllers/MetricExplorerTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/ReportBuilderTest.php [hunk 11]
```diff
@@ -77,6 +78,13 @@ protected function setUp()
             $this->verbose = false;
         }
         $this->helper = new XdmodTestHelper(__DIR__ . '/../../../');
+
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Controllers/ReportBuilderTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/UsageExplorerTest.php [hunk 12]
```diff
@@ -30,13 +31,24 @@ public static function setUpBeforeClass()
     protected function setUp()
     {
         $this->helper = new \TestHarness\XdmodTestHelper();
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Controllers/UsageExplorerTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/UserAdminTest.php [hunk 13]
```diff
@@ -281,9 +282,17 @@ function ($filter) use ($personId) {
      */
     public function provideTestUsersQuickFilters()
     {
-        return Json::loadFile(
-            $this->getTestFiles()->getFile('user_admin', 'user_quick_filters-update_enumAllAvailableRoles', 'output')
-        );
+        $this->setUp();
```
Comment (ryanrath) [inline tests/integration/lib/Controllers/UserAdminTest.php]: `setUp` should be called automatically. Does this fail if you don't explicitly call `setUp`?

File: tests/integration/lib/Database/ResourceNamesTest.php [hunk 14]
```diff
@@ -20,12 +22,80 @@ public function setUp()
     {
         $this->db = DB::factory('datawarehouse');
         $this->testFiles = new TestFiles(__DIR__ . '/../../../');
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Database/ResourceNamesTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Database/ResourceNamesTest.php [hunk 15]
```diff
@@ -20,12 +22,80 @@ public function setUp()
...
+
+        # Check expected file
+        foreach($this->xdmod_realms as $realm) {
+            $expectedOutputFile = $this->testFiles->getFile('integration/database', 'resource_names', "output/$realm");
+
+            # Create missing files/directories
+            if(!is_file($expectedOutputFile)) {
+                $resourceConversions = $this->db->query('SELECT modw.resourcefact.code, modw.resourcetype.abbrev FROM modw.resourcefact
+                INNER JOIN modw.resourcetype ON modw.resourcefact.resourcetype_id=modw.resourcetype.id;');
+                $resource_types = Json::loadFile(__DIR__ . "/../../../../configuration/resource_types.json", false);
```
Comment (ryanrath) [inline tests/integration/lib/Database/ResourceNamesTest.php]: You're going to want to use the `XdmodConfiguration` class as opposed to just loading the raw JSON. This changed recently.

File: tests/integration/lib/Database/SharedJobsTest.php [hunk 16]
```diff
@@ -20,12 +22,80 @@ public function setUp()
     {
         $this->db = DB::factory('datawarehouse');
         $this->testFiles = new TestFiles(__DIR__ . '/../../../');
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Database/SharedJobsTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Database/SharedJobsTest.php [hunk 17]
```diff
@@ -20,12 +22,80 @@ public function setUp()
...
+
+        # Check expected file
+        foreach($this->xdmod_realms as $realm) {
+            $expectedOutputFile = $this->testFiles->getFile('integration/database', 'shared_jobs', "output/$realm");
+
+            # Create missing files/directories
+            if(!is_file($expectedOutputFile)) {
+                $resourceConversions = $this->db->query('SELECT modw.resourcefact.code, modw.resourcetype.abbrev FROM modw.resourcefact
+                INNER JOIN modw.resourcetype ON modw.resourcefact.resourcetype_id=modw.resourcetype.id;');
+                $resource_types = Json::loadFile(__DIR__ . "/../../../../configuration/resource_types.json", false);
```
Comment (ryanrath) [inline tests/integration/lib/Database/SharedJobsTest.php]: *See comment in `ResourceNamesTest`

File: tests/integration/lib/Rest/JobViewerTest.php [hunk 18]
```diff
@@ -10,6 +11,12 @@ public function setUp()
     {
         $xdmodConfig = array( 'decodetextasjson' => true );
         $this->xdmodhelper = new \TestHarness\XdmodTestHelper($xdmodConfig);
+        $xdmod_realms = array();
```
Comment (ryanrath) [inline tests/integration/lib/Rest/JobViewerTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Rest/WarehouseControllerTest.php [hunk 19]
```diff
@@ -1,11 +1,22 @@
 <?php
 
 namespace IntegrationTests\Rest;
+use Models\Services\Realms;
 
 class WarehouseControllerTest extends \PHPUnit_Framework_TestCase
 {
     protected static $helpers = array();
 
+    protected function setUp()
```
Comment (ryanrath) [inline tests/integration/lib/Rest/WarehouseControllerTest.php]: *See comment in `AclsTest`

File: tests/integration/lib/Controllers/UserAdminTest.php (hunk 13 above)
Comment (tpetrus) [reply tests/integration/lib/Controllers/UserAdminTest.php]: Ya, but I'll first try to implement it into BaseTest as you mentioned and see if I still run into the same problem.  For whatever reason, I believe 2 of the tests wouldn't run setUp() before running the test.

File: html/gui/js/modules/job_viewer/JobViewer.js (hunk 2 above)
Comment (tpetrus) [reply html/gui/js/modules/job_viewer/JobViewer.js]: I did this and the Job Viewer tab still shows up if the jobs realm isn't enabled

File: tests/component/lib/SlurmHelperTest.php (hunk 7 above)
Comment (tpetrus) [reply tests/component/lib/SlurmHelperTest.php]: This doesn't seem to work because it doesn't extend BaseTest

File: html/internal_dashboard/js/admin_panel/AclGrid.js [hunk 20]
```diff
@@ -162,7 +162,8 @@ XDMoD.Admin.AclGrid = Ext.extend(Ext.grid.EditorGridPanel, {
 
         // If we are not supporting multiple service providers then pre-fetch
         // the default service provider.
-        if (CCR.xdmod.features.multiple_service_providers === false) {
+        // TODO: Only works with jobs realm enabled
+        if (CCR.xdmod.features.multiple_service_providers === false && XDMOD_REALMS.includes("jobs")) {
```
Comment (ryanrath) [inline html/internal_dashboard/js/admin_panel/AclGrid.js]: You should be able to revert this change and instead update the sql in `XDAdmin::enumerateResourceProviders` to not use `modw_aggregates.jobfact_by_quarter` ( You only need `organization` ).

File: html/gui/js/modules/job_viewer/JobViewer.js (hunk 2 above)
Comment (plessbd) [reply html/gui/js/modules/job_viewer/JobViewer.js]: @tpetrus you are correct this will be resolved in #1006

File: html/internal_dashboard/js/admin_panel/AclGrid.js [hunk 21]
```diff
@@ -162,6 +162,7 @@ XDMoD.Admin.AclGrid = Ext.extend(Ext.grid.EditorGridPanel, {
 
         // If we are not supporting multiple service providers then pre-fetch
         // the default service provider.
+        // TODO: Only works with jobs realm enabled
```
Comment (plessbd) [inline html/internal_dashboard/js/admin_panel/AclGrid.js]: no need for this todo anymore

File: configuration/portal_settings.ini [hunk 22]
```diff
@@ -149,4 +149,4 @@ pass = ""
 database = "mod_hpcdb"
 
 [slurm]
-sacct = "sacct"
+sacct = "sacct"
```
Comment (plessbd) [inline configuration/portal_settings.ini]: put new line back

File: classes/Models/Services/Users.php [hunk 23]
```diff
@@ -118,7 +118,7 @@ public static function getUsers($userTypeId, $aclId, $searchFragment, $requireEm
         if ($searchFragment !== null) {
             $whereClauses [] = <<<SQLF
 (
-  u.username  LIKE CONCAT('%', :filter ,'%') OR 
+  u.username  LIKE CONCAT('%', :filter ,'%') OR
```
Comment (plessbd) [inline classes/Models/Services/Users.php]: no need to make changes to this file.
